### ISSUE TITLE
Pipeline Builds Broken

### ISSUE THREAD

**Issue by ThomasSteinbach** [openshift/jenkins#498]
I recently updated my jenkins-2-centos7 images on my OpenShift cluster with the result, that all Pipeline Builds did not work anymore.

Setup
--------
- I have a simple [build configuration](https://github.com/openshift/jenkins/files/1681212/cd-pipeline.txt)
- ...which is processing my [Jenkinsfile](https://github.com/openshift/jenkins/files/1681192/Jenkinsfile.txt)
- My pipeline is relying on the [OpenShift Jenkins Pipeline (DSL) Plugin](https://github.com/openshift/jenkins-client-plugin)
- The cluster is a RH OpenShift Dedicated installation (see Additional Information below) whose Jenkins does not support this PlugIn
- Therefore I use my own [Jenkins Deployment Configuration](https://github.com/openshift/jenkins/files/1681253/jenkins-dc.txt) using openshift/jenkins-2-centos7

**The pipeline worked** with openshift/jenkins-2-centos7@sha256:4f89fd47a0992710e6a4b2f8634fdd870ad2b01d2fa4095b042133b38ff654e1 created "2 weeks ago"

**The pipeline does not work anymore** with openshift/jenkins-2-centos7@sha256:0adcef7b4cff4f6a4f0a99410099d80d2f2c6071e662535ca03029e6e9b656dc created "4 days ago" 

Now the pipeline execution leads to follwing [error](https://github.com/openshift/jenkins/files/1681286/error.txt)

*Is this enough information to start the investigation of the problem?*

Additional Information
------------------------------
```
$ oc version
oc v3.6.1+008f2d5
kubernetes v1.6.1+5115d708d7
features: Basic-Auth GSSAPI Kerberos SPNEGO

Server https://console.inet-abnahme.ose.db.de:443
openshift v3.6.173.0.21
kubernetes v1.6.1+5115d708d7
```

**Comment by gabemontero** [comment]
hi @ThomasSteinbach 

First, thanks for the detailed set of artifacts.  It very much helped my analysis.  It is rare we get such thorough reporting ;-)

Second, that said, if you ever have future issues you need to report, it would help if you could note two additional points:
a) the version of the openshift client plugin in addition to the image sha, etc.  Though in this case I had to pull the image anyway, but in the future it will help expedite which commit levels to look at wrt the plugin.
b) whether you are running our image in an OpenShift pod, or standalone (i.e. `docker run`); there are many configuration defaults we can pull from the pod that we cannot pull when running standalone, and this can influence diagnosis.

OK, preamble aside, we've seen situations like yours before.  The other moving part to note is the version of pipeline (and its many sub plugins) between the images can change, and what we can and cannot do with declarative pipelines has changed from our perspective as things have evolved.

As such, I'll point you to https://github.com/openshift/jenkins-client-plugin#overview

To summarize, we only support the `openshift.withCluster` directive residing within the `pipeline` directive.  When it is not, we have seen unpredictable results depending on particular pipelines and environments, where situations like you had ... "it used to work, now it does not" occur.

Please try switching the nesting of `openshift.withCluster` and `pipeline` to resolve your issue.
If afterward, something still seems amiss, please reopen / report the details and we'll go from there.

thanks again

@bparees fyi, another one of these nesting of the directive items

**Comment by ThomasSteinbach** [comment]
Cool, I'll try this out as soon as possible (next week).

I also checked my source of information about pipelines, and saw you already updated the [pipeline example](https://github.com/openshift/origin/commit/b5a7a46b47d3a892475821472cf3ef7a86cebca3)

However the original template for my pipeline was the one in the official Origin documentation [here](https://docs.openshift.org/latest/dev_guide/dev_tutorials/openshift_pipeline.html#the-jenkinsfile).

Should I open a new issue or are you able to quickly update this information too?

Thank you so far - Thomas

**Comment by gabemontero** [comment]
On Fri, Feb 2, 2018 at 9:08 AM, Thomas Steinbach <notifications@github.com>
wrote:

> Cool, I'll try this out as soon as possible (next week).
>
> I also checked my source of information about pipelines, and saw you
> already updated the pipeline example
> <https://github.com/openshift/origin/commit/b5a7a46b47d3a892475821472cf3ef7a86cebca3>
>
> However the original template for my pipeline was the one in the official
> Origin documentation here
> <https://docs.openshift.org/latest/dev_guide/dev_tutorials/openshift_pipeline.html#the-jenkinsfile>
>
Should I open a new issue or are you able to quickly update this
> information too?
>
Ah - thanks for the heads up - yes, I should be able to submit a PR against
the documentation later today.
I'll note the PR here when I do.

**Comment by gabemontero** [comment]
On Fri, Feb 2, 2018 at 9:15 AM, Gabe Montero <gmontero@redhat.com> wrote:

>
>
> On Fri, Feb 2, 2018 at 9:08 AM, Thomas Steinbach <notifications@github.com
> > wrote:
>
>> Cool, I'll try this out as soon as possible (next week).
>>
>> I also checked my source of information about pipelines, and saw you
>> already updated the pipeline example
>> <https://github.com/openshift/origin/commit/b5a7a46b47d3a892475821472cf3ef7a86cebca3>
>>
>> However the original template for my pipeline was the one in the official
>> Origin documentation here
>> <https://docs.openshift.org/latest/dev_guide/dev_tutorials/openshift_pipeline.html#the-jenkinsfile>
>>
> Should I open a new issue or are you able to quickly update this
>> information too?
>>
> Ah - thanks for the heads up - yes, I should be able to submit a PR
> against the documentation later today.
> I'll note the PR here when I do.
>

https://github.com/openshift/openshift-docs/pull/7503

**Comment by ThomasSteinbach** [comment]
Hi @gabemontero 
i'll switched the `openshift.withCluster` and `pipeline` statement and - I am sorry to say - things become worse. Now Jenkins complains about a wrong pipeline syntax. To ensure it is not because of my pipeline, I tried the updated nodejs-sample-pipeline above. Here a summary of the setup and result:

* I updated my Jenkins-Deployment to openshift/jenkins-2-centos7@sha256:5fadcb54cdae789b9246248e0cd96e7ebace4fefc3eedbce25160d49634e3838
* This Jenkins (ver. 2.89.3) contains the OpenShift Pipeline Jenkins Plugin in version 1.0.54
* I downloaded your modified [nodejs-sample-pipeline.yaml](https://raw.githubusercontent.com/openshift/origin/b5a7a46b47d3a892475821472cf3ef7a86cebca3/examples/jenkins/pipeline/nodejs-sample-pipeline.yaml)
* ... and imported the BuildConfig via `oc create -f nodejs-sample-pipeline.yaml`

* After starting the pipeline, Jenkins complains about the Jenkinsfile syntax as you can see in the [logs](https://github.com/openshift/jenkins/files/1695405/logs.txt)

-> All images run as pod in OpenShift

I hope I haven't missed something that would lead to this behavior.

**Comment by ThomasSteinbach** [comment]
Because I doubt on myself, here is a second 'script' to reproduce the behavior (easily on Linux):

```
# docker must be installed and running with --insecure-registry 172.30.0.0/16
# oc must be installed

oc cluster up

# wait till the cluster has started
# login with oc as developer

wget -q -O- https://raw.githubusercontent.com/openshift/origin/b5a7a46b47d3a892475821472cf3ef7a86cebca3/examples/jenkins/pipeline/nodejs-sample-pipeline.yaml | oc create -f -

oc start-build nodejs-sample-pipeline

# watch the pipeline build logs in the web gui, and browsing to the jenkins logs
```

I got the same "Method calls on objects not allowed outside "script" blocks..." error messages.

**Comment by gabemontero** [comment]
Thanks for the info @ThomasSteinbach ; I've reopened.  I am still parsing all the details, but at first glance, I believe we are the victim of the seemingly ever changing implementation for pipelines as our image has picked up updates to various dependencies.  The `Method calls on objects not allowed outside "script" blocks.` was *NOT* something we had seen before when we first saw the nesting issue.

I've noticed some recently JIRAs with respect to declarative and global libraries where the technical details are similar to how the `openshift.` directive is actually produced.

I'll report back when I have some more precise details, but at this point, I could see a few things happening:
1) we revise our recommendation on how we coexist with declarative
   a) more work arounds 
   b) or we declare incompatability
2) we possibly revert some pipeline dependencies to work around this, essentially matching the plugin versions from our older image as you noted, though this could have implications wrt security advisories

**Comment by gabemontero** [comment]
@bparees fyi ^^

**Comment by bparees** [comment]
> we possibly revert some pipeline dependencies to work around this, essentially matching the plugin versions from our older image as you noted, though this could have implications wrt security advisories

hopefully we can avoid this one since pinning ourselves on old levels becomes a maintenance problem even outside of security risks

**Comment by gabemontero** [comment]
Ok, I'll direct everyone to https://jenkins.io/doc/book/pipeline/syntax/#script

and to quote that section:  `For most use-cases, the script step should be unnecessary in Declarative Pipelines`

I'm taking the non-decisiveness of `should be` as there caveat for not flinching if they break support, where they then say, "wrap in a script { }` block.  Also note the item about converting large blocks, etc. to global libraries ... interesting in the context of bug reports like https://issues.jenkins-ci.org/browse/JENKINS-46547 ;-)

So ... first stop gap: I got our sample to work like with this:

```
        // path of the template to use
        def templatePath = 'https://raw.githubusercontent.com/openshift/nodejs-ex/master/openshift/templates/nodejs-mongodb.json'
        // name of the template that will be created
        def templateName = 'nodejs-mongodb-example'
        // NOTE, the "pipeline" directive/closure from the declarative pipeline syntax needs to include, or be nested outside,
        // and "openshift" directive/closure from the OpenShift Client Plugin for Jenkins.  Otherwise, the declarative pipeline engine
        // will not be fully engaged.
        pipeline {
            agent {
              node {
                // spin up a node.js slave pod to run this build on
                label 'nodejs'
              }
            }
            options {
                // set a timeout of 20 minutes for this pipeline
                timeout(time: 20, unit: 'MINUTES')
            }

            stages {
                stage('preamble') {
                    steps {
                        script {
                            openshift.withCluster() {
                                openshift.withProject() {
                                    echo "Using project: ${openshift.project()}"
                                }
                            }
                        }
                    }
                }
                stage('cleanup') {
                    steps {
                        script {
                            openshift.withCluster() {
                                openshift.withProject() {
                                    // delete everything with this template label
                                    openshift.selector("all", [ template : templateName ]).delete()
                                    // delete any secrets with this template label
                                    if (openshift.selector("secrets", templateName).exists()) {
                                        openshift.selector("secrets", templateName).delete()
                                    }
                                }
                            }
                        } // script
                    } // steps
                } // stage
                stage('create') {
                    steps {
                        script {
                            openshift.withCluster() {
                                openshift.withProject() {
                                    // create a new application from the templatePath
                                    openshift.newApp(templatePath)
                                }
                            }
                        } // script
                    } // steps
                } // stage
                stage('build') {
                    steps {
                        script {
                            openshift.withCluster() {
                                openshift.withProject() {
                                    def builds = openshift.selector("bc", templateName).related('builds')
                                    // wait up to 5 minutes for the build to complete
                                    timeout(5) {
                                        builds.untilEach(1) {
                                            return (it.object().status.phase == "Complete")
                                        }
                                    }
                                }
                            }
                        } // script
                    } // steps
                } // stage
                stage('deploy') {
                    steps {
                        script {
                            openshift.withCluster() {
                                openshift.withProject() {
                                    def rm = openshift.selector("dc", templateName).rollout()
                                    // wait up to 5 minutes for the deployment to complete
                                    timeout(5) {
                                        openshift.selector("dc", templateName).related('pods').untilEach(1) {
                                            return (it.object().status.phase == "Running")
                                        }
                                    }
                                }
                            }
                        } // script
                    } // steps
                } // stage
                stage('tag') {
                    steps {
                        script {
                            openshift.withCluster() {
                                openshift.withProject() {
                                    // if everything else succeeded, tag the ${templateName}:latest image as ${templateName}-staging:latest
                                    // a pipeline build config for the staging environment can watch for the ${templateName}-staging:latest
                                    // image to change and then deploy it to the staging environment
                                    openshift.tag("${templateName}:latest", "${templateName}-staging:latest")
                                }
                            }
                        } // script
                    } // steps
                } // stage
            } // stages
        } // pipeline

```

Undoubtedly tedious in comparison to what the declarative syntax parsing engine used to allow.

Updating our sample, and updating the doc to require this approach, will be our fallback.  As @bparees noted, pining ourselves to older, more permissive declarative pipeline versions is untenable for any prolonged duration.

Before giving up though, I'm going to dive into the declarative pipeline engine code again, and see if there is some what we can mark our plugin so that `script { }` is assumed or not required.

**Comment by gabemontero** [comment]
Still looking, but found a clue in their tests that produce the `Method calls on objects not allowed outside "script" blocks` message ... the test case is suppose to fail because:

```
Passing a body that isn't composed of steps does not work.
```

Certainly our `openshift.` closures fall in that category.

**Comment by gabemontero** [comment]
Looking at the [docker workflow plugin](https://github.com/jenkinsci/docker-workflow-plugin) (whose global variable based design motivated this plugin), discussion of integration with declarative is minimal, other than being able to control / specify agents.  And the samples are all scripted pipeline.

**Comment by ThomasSteinbach** [comment]
FYI - I tried the _scripted_ Jenkins pipeline which worked fine for me. I know this is not the ultimate solution. Also because the Jenkins scripted pipeline is poorly documented. But it is an alternative way to work with the OpenShift Jenkins Client Plugin. Here the sample pipeline 'scripted':

```
def templatePath = 'https://raw.githubusercontent.com/openshift/nodejs-ex/master/openshift/templates/nodejs-mongodb.json'
def templateName = 'nodejs-mongodb-example'

openshift.withCluster() {
  node {
        stage('cleanup') {
            openshift.selector("all", [ template : templateName ]).delete()
            if (openshift.selector("secrets", templateName).exists()) {
                openshift.selector("secrets", templateName).delete()
            }
        }

        stage('create') {
            // create a new application from the templatePath
            openshift.newApp(templatePath)
        }
        
        stage('build') {
            def builds = openshift.selector("bc", templateName).related('builds')
            timeout(5) {
                builds.untilEach(1) {
                    return (it.object().status.phase == "Complete")
                }
            }
        }
          
        stage('deploy') {
            def rm = openshift.selector("dc", templateName).rollout()
            timeout(5) {
                openshift.selector("dc", templateName).related('pods').untilEach(1) {
                 return (it.object().status.phase == "Running")
                }
            }
        }
          
        stage('tag') {
            openshift.tag("${templateName}:latest", "${templateName}-staging:latest")
         }
    }
}
```

**Comment by gabemontero** [comment]
On Tue, Feb 6, 2018 at 7:39 AM, Thomas Steinbach <notifications@github.com>
wrote:

> FYI - I tried the *scripted* Jenkins pipeline which worked fine for me. I
> know this is not the ultimate solution. Also because the Jenkins scripted
> pipeline is poorly documented. But it is an alternative way to work with
> the OpenShift Jenkins Client Plugin. Here the sample pipeline 'scripted':
>
> def templatePath = 'https://raw.githubusercontent.com/openshift/nodejs-ex/master/openshift/templates/nodejs-mongodb.json'
> def templateName = 'nodejs-mongodb-example'
>
> openshift.withCluster() {
>   node {
>         stage('cleanup') {
>             openshift.selector("all", [ template : templateName ]).delete()
>             if (openshift.selector("secrets", templateName).exists()) {
>                 openshift.selector("secrets", templateName).delete()
>             }
>         }
>
>         stage('create') {
>             // create a new application from the templatePath
>             openshift.newApp(templatePath)
>         }
>
>         stage('build') {
>             def builds = openshift.selector("bc", templateName).related('builds')
>             timeout(5) {
>                 builds.untilEach(1) {
>                     return (it.object().status.phase == "Complete")
>                 }
>             }
>         }
>
>         stage('deploy') {
>             def rm = openshift.selector("dc", templateName).rollout()
>             timeout(5) {
>                 openshift.selector("dc", templateName).related('pods').untilEach(1) {
>                  return (it.object().status.phase == "Running")
>                 }
>             }
>         }
>
>         stage('tag') {
>             openshift.tag("${templateName}:latest", "${templateName}-staging:latest")
>          }
>     }
> }
>
> Yep this totally has to do with declarative.  This example was originally
just scripted, but we
added the use of declarative as a litmus test and show how they could
coexist, given the clear
indication that declarative is the strategic direction with upstream
Jenkins.  But as I mentioned
before declarative is new and obviously is rapidly evolving.

I have a few more investigative threads I hope to complete today, and then
make a final decision
on this, where we'll have to continually monitor things and see when /
where we can pivot.

thanks

**Comment by gabemontero** [comment]
So one experiment to analyze my "global variable with bodies" not supported, vs. say a single "script step" being supported, provided confirmation.

This declarative pipeline using our legacy plugin's pipeline support is accepted:

```
pipeline {
    agent any
    options {
        // set a timeout of 20 minutes for this pipeline
        timeout(time: 20, unit: 'MINUTES')
    }

    stages {
        stage('verify') {
            steps {
                openshiftVerifyService apiURL: '', authToken: '', namespace: '', svcName: 'jenkins', verbose: 'false'
            }
        }
    } // stages
} // pipeline
```

Note I did not have to wrapper `openshiftVerifyService...` with `script {...}`.

**Comment by gabemontero** [comment]
And sure enough, the cloudbees docker workflow plugin exhibits the same behavior:

```
WorkflowScript: 11: Method calls on objects not allowed outside "script" blocks. @ line 11, column 17.
                   docker.withRegistry('https://docker.io') {
```

**Comment by gabemontero** [comment]
OK, last bullet ... I've spent a few hours seeing how we might be able to leverage org.jenkinsci.plugins.pipeline.modeldefinition.ModelInterpreter.groovy

this is the Global Variable for the declarative pipeline support ... the analogous piece to this plugin's OpenShiftDSL.groovy; it is the entry point for `pipeline { ... }`

no luck

there is some interesting discussion in the https://github.com/jenkinsci/pipeline-model-definition-plugin/blob/master/DEV_GUIDE.md about how ModelInterpreter works, and that they are monitoring some groovy related items, and things might evolve again

So, for the time being, we are falling back to requiring the pattern I specified in https://github.com/openshift/jenkins/issues/498#issuecomment-363220347 if you want to leverage declarative syntax; I'll use this issue to drive the update to this plugin's readme, and I'll cross reference the openshift/origin PR and openshfit/openshift-docs PR updating our sample with this issue.

**Comment by ThomasSteinbach** [comment]
Thank you for your effort and open discussion. It showed me to use the OpenShift Jenkins Client Plugin with scripted pipelines as I mentioned [above](https://github.com/openshift/jenkins/issues/498#issuecomment-363410015), which is totally fine for me right now.

**Comment by gabemontero** [comment]
great / glad to help @ThomasSteinbach 

and thanks again for the heads up on what was going on ... as a side effect off all this, I'm making sure the nodejs-sample-pipeline is part of our nightly regression :-)
